### PR TITLE
Contact config - message expiration time

### PR DESCRIPTION
**Pull request by matuskosut**
Could this be a way how to approach setting of expiration time for messages (on the contact level) through the client? As requested for example by https://github.com/AsamK/signal-cli/issues/153 and it may also help moving on with this one too https://github.com/AsamK/signal-cli/issues/291

If yes, I could add group configuration part or asses the current one if you have any feedback.

### CODE REVIEW

File: src/main/java/org/asamk/signal/commands/UpdateContactCommand.java [hunk 1]
```diff
@@ -15,6 +15,10 @@ public void attachToSubparser(final Subparser subparser) {
         subparser.addArgument("-n", "--name")
                 .required(true)
                 .help("New contact name");
+        subparser.addArgument("-e", "--expiration")
+                .required(false)
+                .type(int.class)
+                .help("Specify expiration time of messages");
```
Comment (AsamK) [inline src/main/java/org/asamk/signal/commands/UpdateContactCommand.java:21]: Please mention the unit here (seconds) and also document the argument in the `man/signal-cli.1.adoc` man page file.

File: src/main/java/org/asamk/signal/commands/UpdateContactCommand.java [hunk 2]
```diff
@@ -30,6 +34,12 @@ public int handleCommand(final Namespace ns, final Manager m) {
 
         try {
             m.setContactName(number, name);
+
+            int expiration = 0;
```
Comment (AsamK) [inline src/main/java/org/asamk/signal/commands/UpdateContactCommand.java:38]: you can make `expiration` an `Integer` and already assign it here, like this:
```java
            Integer expiration = ns.getInt("expiration");
            if (expiration != null) {
                m.setExpirationTimer(number, expiration);
            }

File: src/main/java/org/asamk/signal/manager/Manager.java [hunk 3]
```diff
@@ -810,10 +810,12 @@ public String getGroupName(byte[] groupId) {
     /**
      * Change the expiration timer for a contact
      */
-    public void setExpirationTimer(SignalServiceAddress address, int messageExpirationTimer) {
```
Comment (AsamK) [inline src/main/java/org/asamk/signal/manager/Manager.java:813]: please keep the method with `SignalServiceAddress` parameter and just add a new method (with the same name) that takes a `String` and does the `canonicalizeAndResolveSignalServiceAddress` conversion.

File: src/main/java/org/asamk/signal/manager/Manager.java [hunk 4]
```diff
@@ -810,10 +810,12 @@ public String getGroupName(byte[] groupId) {
...
-        ContactInfo c = account.getContactStore().getContact(address);
+    public void setExpirationTimer(String number, int messageExpirationTimer) throws InvalidNumberException {
+        ContactInfo c = account.getContactStore().getContact(canonicalizeAndResolveSignalServiceAddress(number));
         c.messageExpirationTime = messageExpirationTimer;
+        System.err.println("Set message expiration " + c.number + " -> " + c.messageExpirationTime + " seconds");
```
Comment (AsamK) [inline src/main/java/org/asamk/signal/manager/Manager.java:816]: I'd like to get rid of all console output in the Manager class, so please remove the println here as well.
As long as a command completes successfully, I'd prefer to not output anything as long as it doesn't provide additional information.

Comment (matuskosut) [reply src/main/java/org/asamk/signal/manager/Manager.java]: Sure this makes sense, removed it

File: src/main/java/org/asamk/signal/manager/Manager.java (hunk 3 above)
Comment (matuskosut) [reply src/main/java/org/asamk/signal/manager/Manager.java]: edited, you can review the changes

**Comment by matuskosut** [comment]
worked through the small things, good to review. When testing it sets the expiration correctly when first message after is sent, but I have to look a bit more on how to notify when it is set, and to make sure other devices get the change

File: man/signal-cli.1.adoc [hunk 5]
```diff
@@ -252,6 +252,10 @@ Specify the contact phone number.
 *-n*, *--name*::
 Specify the new name for this contact.
 
+*-e*, *--expiration*::
+Set expiration time of messages (seconds).
+By default expiration is set to 0, no expiration.
```
Comment (AsamK) [inline man/signal-cli.1.adoc:257]: hm, this is a bit misleading, when adding a new contact, the expiration is set to 0 by default, but when updating a contact the expiration is left unchanged by default.

**Comment by AsamK** [comment]
To notify the other contact, it should be enough to send a message with an expiration update:
```java
messageBuilder.withExpiration(contact.messageExpirationTime);
messageBuilder.asExpirationUpdate();
```
Sending it to the other contact will also trigger a sync message to the linked devices.

File: man/signal-cli.1.adoc (hunk 5 above)
Comment (matuskosut) [reply man/signal-cli.1.adoc]: I changed the second part to `To disable expiration set expiration time to 0.`, does it feel better? Or would you remove that part completely?

**Comment by matuskosut** [comment]
thanks for the hints, I added message sending

**Pull request merged**